File: 0048-rotate-image/0048-rotate-image.js
/**
 * @param {number[][]} matrix
 * @return {void} Do not return anything, modify matrix in-place instead.
 */
function transpose(matrix) {
    const n = matrix.length;
    for (let i = 0; i < n; i++) {
        for (let j = i; j < n; j++) {
            let temp = matrix[i][j];
            matrix[i][j] = matrix[j][i];
            matrix[j][i] = temp;
        }
    }
}
var rotate = function (matrix) {


    transpose(matrix);
    const n = matrix.length;
    for (let i = 0; i < n; i++) {
        let start = 0;
        let end = n - 1;
        while (start < end) {
            let temp = matrix[i][start];
            matrix[i][start] = matrix[i][end];
            matrix[i][end] = temp;
            start++;
            end--;
        }
    }


};

File: 0048-rotate-image/0048-rotate-image.cpp
class Solution {
public:
    void transpose(vector<vector<int>>& matrix, int n){
        for(int i=0;i<n;i++){
            for(int j=i;j<n;j++){
                swap(matrix[i][j], matrix[j][i]);
            }
        }
    }
    void rotate(vector<vector<int>>& matrix) {
        int n = matrix.size();

        transpose(matrix, n);

        for(int i=0;i<n;i++){
            int l=0;
            int r= n-1;
            while(r>l){
                swap(matrix[i][l], matrix[i][r]);
                l++;
                r--;
            }
        }
        
    }
};